src/js/DOMUtils.js

```diff
@@ -0,0 +1,197 @@
+        if (pathNode.localName == 'content')
+            continue;
+        if (pathNode.nodeType === Node.DOCUMENT_FRAGMENT_NODE)
+            continue;
+        return pathNode;
+    }
+    return null;
+};
+})();
+
+/**
+ * Return the corresponding element for the given node.
+ * @param {Node} node
+ * @return {Element}
+ * @suppress {checkTypes}
+ */
+axs.dom.asElement = function(node) {
+    /** @type {Element} */ var element;
+    switch (node.nodeType) {
+    case Node.COMMENT_NODE:
+        return null;  // Skip comments
+    case Node.ELEMENT_NODE:
+        element = /** (@type {Element}) */ node;
+        if (element.localName == 'script')
+            return null;  // Skip script elements
+        break;
+    case Node.TEXT_NODE:
+        element = axs.dom.parentElement(node);
+        break;
+    default:
+        console.warn('Unhandled node type: ', node.nodeType);
+        return null;
+    }
+    return element;
+};
+
+/**
+ * Recursively walk the composed tree from |node|, aborting if |end| is encountered.
+ * @param {Node} node
+ * @param {?Node} end
+ * @param {{preorder: (function (Node)|undefined),
+ *          postorder: (function (Node)|undefined)}} callbacks
+ *     Callbacks to be called for each element traversed, excluding
+ *     |end|. Possible callbacks are |preorder|, called before descending into
+ *     child nodes, and |postorder| called after all child nodes have been
+ *     traversed.
+ * @param {ShadowRoot=} opt_shadowRoot The nearest ShadowRoot ancestor, if any.
+ * @return {boolean} Whether |end| was found, if provided.
+ */
+axs.dom.composedTreeSearch = function(node, end, callbacks, opt_shadowRoot) {
+    if (node === end)
+        return true;
+
+    if (node.nodeType == Node.ELEMENT_NODE)
+        var element = /** @type {Element} */ (node);
+
+    if (element && callbacks.preorder)
+        callbacks.preorder.call(null, element);
+
+    var found = false;
+
+    // Descend into node:
+    // If it has a ShadowRoot, ignore all child elements - these will be picked
+    // up by the <content> or <shadow> elements. Descend straight into the
+    // ShadowRoot.
+    if (element) {
+        // NOTE: grunt qunit DOES NOT support Shadow DOM, so if changing this
+        // code, be sure to run the tests in the browser before committing.
+        var shadowRoot = element.shadowRoot || element.webkitShadowRoot;
+        if (shadowRoot) {
+            found = axs.dom.composedTreeSearch(shadowRoot,
+                                               end,
+                                               callbacks,
+                                               shadowRoot);
+            if (element && callbacks.postorder && !found)
+                callbacks.postorder.call(null, element);
+            return found;
+        }
+    }
+
+    // If it is a <content> element, descend into distributed elements - descend
+    // into distributed elements - these are elements from outside the shadow
```

Comment:
Did you double up typing `descend into distributed elements` intentionally here?